### ISSUE TITLE
[question] how to debug apparently-unnecessary rebuilds?

### ISSUE THREAD

**Issue by gasche** [ocaml/dune#4150]
I am creating a Gitlab CI script for a project using Dune. This involves caching things and reusing them on following runs. If I ask to cache Dune's `_build` directory, and start a new CI job without any change to the source, then *some* of the build will get reused as expected, but some other parts of the build get redone for no clear reason.

Question: is there a way to debug apparently-unnecessary rebuilds? For example it could be a `--display` option that, for each action that is performed, explains why the action was deemed necessary. Something more low-level would be fine, but preferably this would not require modifying the Dune implementation.

While trying to answer this question I found `dune rules`. This is useful, but it show all dependencies for each rule, and I don't know which of the dependencies are causing the rebuild. How would I go about checking that?

**Comment by gasche** [comment]
I suspect that the issue comes from an "external" dependency. Is there an explanation somewhere of how Dune checks that external dependencies are up-to-date? (Ideally there would be a way to check what it knows about them in its cache, and to compute the same information for a given external path.)

**Comment by rgrinberg** [comment]
>  Is there an explanation somewhere of how Dune checks that external dependencies are up-to-date?

External dependencies are not treated in a special way. If a binary or a file is external to the workspace, dune simply records a dependency on it.

**Comment by gasche** [comment]
Then I guess the [second question](https://github.com/ocaml/dune/issues/4150#issuecomment-766248411) is: how can we know whether Dune will consider a dependency changed or not? For example, if they have the same `sha256sum`, can it still cause a rebuild? (Is this documented somewhere?)

**Comment by gasche** [comment]
Note: my [gitlab-ocaml-ci-example](https://gitlab.com/gasche/gitlab-ocaml-ci-example/) repository is an example of this caching issue on Gitlab. [This CI build](https://gitlab.com/gasche/gitlab-ocaml-ci-example/-/jobs/982605491) is an example where nothing in the OCaml source changed. One can see that dune performed no actions in the "run the tests" and "build the documentation" steps, but it did rebuild some objects during the "build the project" steps (but not all of them, `.d` files were not rebuilt for example).

**Comment by nojb** [comment]
@gasche This is not a direct answer to your question, but have you considered using `dune cache` https://dune.readthedocs.io/en/stable/caching.html? I seem to remember this worked better than blindly copying the `_build` directory.

**Comment by gasche** [comment]
I did not use `dune cache` for several reasons:
- My understanding is that the main point of the dune cache is to share work among distinct builds, while my CI use-case is to repeatedly compiling the same project in the same environment, with small code changes. Naively this use-case sounds like a good fit for `_build`, with little benefits from a global cache. (Having a longer memory can help when testing several branches, but this is marginal especially for simple proejcts.)
- Gitlab CI only lets you cache subdirectories of the current build directory, while the Dune global cache lives outside the directory (but `_build`is right here), so it would require a cumbersome dance of moving things around.

**Comment by jeremiedimino** [comment]
I guess we could add an option for Dune to print files that it checked for modifications and found different. Combined with printing dependency stacks, that would be a cheap way to debug spurious rebuilds.

> how can we know whether Dune will consider a dependency changed or not? For example, if they have the same sha256sum, can it still cause a rebuild? (Is this documented somewhere?)

Dune looks at the md5 of `<md5-of-file-contents><executable-bit-as-one-byte>`. If that changes, the file is considered as changed. This is not documented, and subject to change in the future.

**Comment by gasche** [comment]
Thanks! (Yes, I would welcome such a rule-activation-explaining mode.)

Do I understand correctly that, currently, the modification time of things is not taken into account? I did and am still doing various experiments with `touch`, this should not affect the result?

Of course the way the digest is computed needs not be documented, but maybe it would be reasonable to document that you are looking at the file content and the executable bit? Or do you want to keep the ability to add more change-criteria in the future and you think that documenting the current ones would prevent that?

**Comment by ejgallego** [comment]
@gasche a common source of problems here is Gitlab's CI not preserving executable bits in `_build` due to the way artifacts are compressed, see for example https://github.com/coq/coq/issues/10694 where we hit this bug.

To debug this I usually modify the digest function to output a message when a digest has changed, as outlined in the bug above, dune will include the `x` bit in the digest so Gitlab's `_build` directory will trigger lots of spurious rebuilds.

**Comment by ejgallego** [comment]
> Do I understand correctly that, currently, the modification time of things is not taken into account? I did and am still doing various experiments with `touch`, this should not affect the result?

As discussed in the Coq-related PR https://github.com/coq/coq/pull/10695 , dune's digest rule is:
```ocaml
Digest.generic (file_contents f, f.stats.st_perm land 0o100 (* Only take USR_X in account *))
```
so it is the md5 + x bit.

**Comment by ejgallego** [comment]
IMHO when a file is considered updated could be a bit better documented, along with common pitfalls such as this permission problem.

**Comment by ejgallego** [comment]
> ```ocaml
> Digest.generic (file_contents f, f.stats.st_perm land 0o100 (* Only take USR_X in account *))
> ```

Code has quite changed since that version, the relevant bits are now in `src/dune_engine/cached_digest.ml` and `Digest.file_with_stats` , sorry for the stale info.

**Comment by gasche** [comment]
Thanks a lot @ejgallego, this looks like it is exactly the issue.

Would you by chance volunteer to implement the "explain what dependency has changed  to cause the rule to run" log/display mode, so that the next person after you and myself does not need to spend as much time trying to figure out the issue?

**Comment by ejgallego** [comment]
> Would you by chance volunteer to implement the "explain what dependency has changed to cause the rule to run" log/display mode, so that the next person after you and myself does not need to spend as much time trying to figure out the issue?

I'd love to but unfortunately my backlog is just too large these days, we had already some discussion about implementing this kind of debug mode [along with cutoffs so one can tell dune to ignore a changed digest] , so IMHO we could just summarize the TODO in a issue so anyone could take on this task. We need to verify if an issue requesting this doesn't yet exist.

**Comment by gasche** [comment]
I think that the following would be really useful:

- Some documentation of what aspects of files are currently considered to decide to rebuild.
- A `--display=changes` mode that is like `--display=short`, but additionally gives a justification for each rule that is being run.

I could help with the former (but it sounds like @jeremiedimino may have better ideas on what should or maybe should not be documented), but for the latter it would be much easier to do for someone with previous experience on the Dune codebase, so I would rather leave that to such a volunteer.

**Comment by jeremiedimino** [comment]
> Do I understand correctly that, currently, the modification time of things is not taken into account? I did and am still doing various experiments with touch, this should not affect the result?

That's right. However, computing the digest of the contents is expensive, so this computation itself is cached. To decide whether a file should be re-hashed, Dune looks at the mtime (and a few other stats). This of course has some pitfalls; if your file-system clock has a high granularity, then Dune will re-hash files more often than necessary. FTR, we have encountered file systems that store mtimes with a 1s granularity.

> Of course the way the digest is computed needs not be documented, but maybe it would be reasonable to document that you are looking at the file content and the executable bit? Or do you want to keep the ability to add more change-criteria in the future and you think that documenting the current ones would prevent that?

Documenting it is fine as long as:
- it is clear that they is no stability promise
- the text doesn't go out of date the next time we touch this code
So the text should be a bit loose. For instance: "To determine if a file has changed, Dune remembers a few facts about the file. For instance a digest of the contents and a few stats like the executable bit".

**Comment by gasche** [comment]
Note: while I do believe that the issue I observe is exactly the same as @ejgallego (my logging shows that files in the `_build` cache became executable for no obvious reason), I am presently unable to reproduce the issue in a simplified environment (... while the repository I'm starting from is already fairly simple). (In particular, in all my simplified-reproduction-tests so far Gitlab correctly restores files permissions.)

**Comment by ejgallego** [comment]
@gasche maybe an issue with symlinks and `.zip` files?

**Comment by gasche** [comment]
I have managed to produce a minimal test-case for this bug, which definitely is on the Gitlab side, and yes involves symlinks. (Their unzipper code sets permissions `rwxrwxrwx` when unpacking any file that is the *target* of a symlink.) Upstream [issue report](https://gitlab.com/gitlab-org/gitlab-runner/-/issues/27496).

Note that while my particular problem is solved, the issue that I don't know how to effectively debug apparently-unnecessary rebuilds in `dune` is still open.

**Comment by jeremiedimino** [comment]
Agreed. We are currently in a code-sprint to switch from Jenga to Dune inside Jane Street. Once we use Dune inside Jane Street, I expect that eventually we'll want to debug similar cases and we'll probably add the necessary features at this point if they haven't been added already.

**Comment by ejgallego** [comment]
As I had to debug this kind of problems a couple of times, let me add a couple of comments in case they would be helpful:
- two main worries that scared me from trying to upstream my debug patches were efficiency and output. This kind of debug code sits in a hot path so it could be delicate, also how to do a reasonable interface is not easy, due to verbosity and code flow [see below]
- a very simple way to add some debug code is to modify the digest code as to complain when a digest of file has been changed, however this is hacky and doesn't capture other reasons
- a more principled way would be to hook in 
https://github.com/ocaml/dune/blob/f0a31ae77f5f161e52a37548e6754987fd9b44ae/src/dune_engine/build_system.ml#L1382-L1402 , unfortunately at this point the trace of the rule has been computed, so you would need to rerun it with a "diff" algorithm to actually see what part of the rule is to blame for the different trace; I guess that is not so complicated, other than this being kind of critical code so I was a bit scared to mess with it.
